packages/workflow/src/workflow.ts

```diff
@@ -1122,11 +1122,11 @@ function conditionInner(fn: () => boolean): Promise<void> {
  * Definitions are used to register handler in the Workflow via {@link setHandler} and to signal Workflows using a {@link WorkflowHandle}, {@link ChildWorkflowHandle} or {@link ExternalWorkflowHandle}.
  * Definitions can be reused in multiple Workflows.
  */
-export function defineSignal<Args extends any[] = []>(name: string): SignalDefinition<Args> {
+export function defineSignal<Args extends any[] = [], Name extends string = string>(name: Name): SignalDefinition<Args, Name> {
```

Comment:
I don't love this extra type arg since it cannot be inferred from the function's parameter and requires duplicate definition.

Some alternatives, by least to most favorite:

```ts
export function defineSignalUglyVersion<Args extends any[] = any[], Name extends string = string>(
  name: Name,
  args?: Args
): SignalDefinition<Args, Name> {
  return {} as any;
}

export function namedSignalDefiner<Name extends string>(name: Name): Definer<Name> {
  return {} as any;
}

defineSignal<[string]>('a');
defineSignalUglyVersion('a', {} as [string]);
defineSignalUglyVersion('a');

interface Definer<Name extends string> {
  define<Args extends any[]>(): SignalDefinition<Args, Name>;
}

const s = namedSignalDefiner('a').define<[string]>();

// Another version of this can rewrite the returned object keys to be `signal${K}` but that's probably overkill.
export type SignalsForDefs<T> = {
  [K in keyof T & string]: T[K] extends (...args: infer A) => void ? SignalDefinition<A, K> : never;
};

export function defineSignals<Defs>(): SignalsForDefs<Defs> {
  return {} as any;
}

interface MySignals {
  a(): void;
}

const signals = defineSignals<MySignals>();
```

WDYT?

Reply:
Other variations of these solutions are possible too.

@mjameswh your input is also welcome.

Reply:
Combining/extending some of @bergundy ideas by using overloads on `defineSignal`:

```
interface SignalDefinitionWithUnspecifiedArgs<Name extends string> extends SignalDefinition<[], Name> {
  withArgs<Args extends any[]>(): SignalDefinition<Args, Name>;
}

// defineSignal('signalA')
//  - Return value extends SignalDefinition<[], Name> which maintain backward compatibility
//  - Return value allows refining arg types using .withArgs()
//
export function defineSignal<Name extends string>(name: Name): SignalDefinitionWithUnspecifiedArgs<Name>;

// defineSignal('signalB', mySignalHandlerFunction)
//  - deduce signal args from the handler function
//  - technically nice, but probably rarely applicable in practice, as handlers are generally not in scope at the moment `defineSignal` is called (ie. handlers are generally defined as inner functions inside the workflow function itself)
//  - Can be worked around by creating an anonymous noop function just for the purpose of `defineSignal`, but that is no better than withArgs<[ ... ]>()
//
export function defineSignal<Args extends any[], Name extends string>(
  name: Name,
  fn: (...args: Args) => void | Promise<void>
): SignalDefinition<Args, Name>;

// defineSignal<[number, string]>('signalC')
//   - backward compatible form (note that in this case, Name is string)
//
// defineSignal<[number, string], 'signalC'>('signalC')
//   - Explicit format, where both Args and Name are specified
//
export function defineSignal<Args extends any[], Name extends string = string>(
  name: Name
): SignalDefinition<Args, Name>;

// Implementation
export function defineSignal<Args extends any[] = [], Name extends string = string>(
  name: Name
): SignalDefinitionWithUnspecifiedArgs<Name> | SignalDefinition<Args, Name> {
  return {
    type: 'signal',
    name,
  } as SignalDefinitionWithUnspecifiedArgs<Args, Name>;
}

// Example usage
const signalFoo = defineSignal('foo').withArgs<number, string>();
```

Reply:
As a side note, we have two features on our road map that might also require revisiting the way signal handlers are defined:

- Support for signal/query introspection (ie. allow for the web UI to present a list of signals and queries supported by a given workflow, including the number and type of arguments)
- Make it possible to register multiple signal handlers at once, so that buffered signals get dispatch in the right order.

If we are to actually change the way signals are defined, I think we should at least consider either the new API actually get us closer from these two goals (don't mean we should _implement_ those features immediately). It would be disappointing that we encourage users to change their code now, and then come up with a completely different recommandation in three months...

Reply:
In the meantime, I suggest we merge this since it's only improving the existing situation and consider a better API later on.

Reply:
Sorry I haven't been able to get back to these considerations - great conversation here around a lot of ergonomics options. As @bergundy mentioned, I've intentionally written this PR to be zero-impact on existing callers (other than type safety), and that does mean some compromises.

> Support for signal/query introspection (ie. allow for the web UI to present a list of signals and queries supported by a given workflow, including the number and type of arguments)

This would be fantastic; my team is about to implement a workflow abstraction layer to do this ourselves, based on the changes in this PR. But first-class support would of course be better.